compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -250,14 +376,20 @@ pub fn push_debuginfo_type_name<'tcx>(
         output: &mut String,
         visited: &mut FxHashSet<Ty<'tcx>>,
     ) {
-        let layout = tcx.layout_of(tcx.param_env(def.did).and(ty)).expect("layout error");
-
-        if let Variants::Multiple {
-            tag_encoding: TagEncoding::Niche { dataful_variant, .. },
-            tag,
-            variants,
+        if let Ok(TyAndLayout {
+            layout:
+                Layout {
+                    variants:
+                        Variants::Multiple {
+                            tag_encoding: TagEncoding::Niche { dataful_variant, .. },
+                            tag,
+                            variants,
+                            ..
+                        },
+                    ..
+                },
             ..
-        } = &layout.variants
+        }) = &tcx.layout_of(tcx.param_env(def.did).and(ty))
```

Comment:
@wesleywiser, could you take a look at this change?

Reply:
This seems ok to me but we really shouldn't be getting an `Err(...)` from `layout_of` at this point since we've run the rest of the compiler and haven't encountered any errors. @dpaoliello did you run into an ICE or something that caused you to change this or was this just a stylistic change?

Reply:
I hit an ICE, specifically for open generic enums (for example when running tests it ICEs for `std::result::Result<T, F>`).

I think we're only hitting this with my change as implementations print out the open generic in the name:
```
a!function_names::impl$<function_names::TestTrait1, GenericStruct<T, i32> >::trait_function<i32>
```

Reply:
Can you post a repro? I'm a bit concerned this is an issue in my changes.

Reply:
Easiest way to repro is to:
1. Pull my branch
2. Add back in the `expect` call
3. Run the debuginfo tests

Reply:
Got it. My concern was that we might be covering up some path where we're operating on a polymorphic type but should be operating on a monomorphic one. I checked out your branch and I'm not concerned about that anymore and see why we need to do this. 

Thanks!